### PR TITLE
RUMM-1153: Possibility to attach custom attributes to automatically collected RUM Resource events

### PR DESCRIPTION
**Pull request by 0xnm**
### What does this PR do?

This change gives user a possibility to attach custom attributes to automatically collected RUM Resource events: either in the `OkHttp` flow, or in the `WebView` flow.

Attributes are collected after the resource call is completed (with either `success` or `error` outcome)

### Additional Notes

Resources are also automatically collected [here](https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/rum/resource/RumResourceInputStream.kt), but this part is not covered by this PR, because this class is low level and not sure if it is needed. Can be addressed in another PR if needed though.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [x] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt [hunk 1]
```diff
@@ -65,13 +67,16 @@ import okhttp3.Response
...
- *
+ * @param rumResourceAttributesProvider which listens on the intercepted [okhttp3.Request]
+ * and offers the possibility to add custom attributes to the RUM resource events.
  */
 open class DatadogInterceptor
 internal constructor(
     tracedHosts: List<String>,
     tracedRequestListener: TracedRequestListener,
     firstPartyHostDetector: FirstPartyHostDetector,
+    private val rumResourceAttributesProvider: RumResourceAttributesProvider =
```
Comment (0xnm) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt:78]: it is not last in the argument list to keep lambda last to still have `ctor() { lambda }` syntax

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 2]
```diff
@@ -281,11 +283,35 @@ class NoOpFactoryAnnotationProcessor : AbstractProcessor() {
                 funSpecBuilder.addStatement("return %T.${firstValue.simpleName}", type)
             }
             returnTypeDef.kind == ElementKind.INTERFACE -> {
-                val packageName = (returnTypeDef as TypeElement).qualifiedName
-                    .toString()
-                    .substringBeforeLast('.')
-                val noOpReturnType = ClassName(packageName, "NoOp${returnTypeDef.simpleName}")
-                funSpecBuilder.addStatement("return %T()", noOpReturnType)
+                val returnTypeElement = returnTypeDef as TypeElement
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:286]: Nicely done

File: sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt [hunk 3]
```diff
@@ -50,7 +53,16 @@ class SampleApplication : Application() {
...
     private val okHttpClient = OkHttpClient.Builder()
-        .addInterceptor(RumInterceptor())
+        .addInterceptor(
+            RumInterceptor(
+                rumResourceAttributesProvider = object : RumResourceAttributesProvider {
+                    override fun onProvideAttributes(
+                        request: Request,
+                        response: Response?,
+                        throwable: Throwable?
+                    ): Map<String, String> = mapOf("foo" to "bar")
```
Comment (mariusc83) [inline sample/kotlin/src/main/kotlin/com/datadog/android/sample/SampleApplication.kt:63]: hmmm...I'd rather test it if it works but not add those in the sample app. A good place where you can add this and you could do this in a next task is in the `integration tests` project and try to assert for them in the mock server results.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogInterceptorTest.kt [hunk 4]
```diff
@@ -112,6 +142,12 @@ internal class DatadogInterceptorTest : TracingInterceptorNotSendingSpanTest() {
                 expectedStopAttrs
             )
         }
+
+        verify(mockRumAttributesProvider, times(1)).onProvideAttributes(
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogInterceptorTest.kt:146]: not sure you need to test this here. You are already asserting the attributes above.

Comment (0xnm) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogInterceptorTest.kt:146]: agree, probably will remove these assertions as long as attributes are verified

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import okhttp3.Response
+import org.junit.Assert.assertThrows
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+internal class ResourceTrackingTest {
```
Comment (mariusc83) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt:34]: very nice !!

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    private fun List<HandledRequest>.rumEventsBy(
+        predicate: (rawEvent: JsonObject) -> Boolean
+    ): List<JsonObject> {
+        return filter { it.url?.isRumUrl() ?: false }
+            .flatMap { rumPayloadToJsonList(it.textBody!!) }
+            .filter { predicate.invoke(it) }
+    }
+
+    private fun Forge.exhaustiveAttributes(): Map<String, Any?> {
```
Comment (mariusc83) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt:135]: maybe we should extract this in a `ForgeExt.kt` file. I am pretty sure we will need it in some other places also.

**Review by mariusc83 — APPROVED** [review]

**Review by xgouchet — APPROVED** [review]

**Pull request merged**